File: RB101_RB109_small_problems/medium_2/next_featured_number_higher_than_a_given_value.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: single integer
-  Output: integer representing the next featured number greater than the argument OR an error if there is no next featured number

---

**Problem Domain:**

Featured number is an odd number that:
  - is an odd number
  - is a multiple of 7
  - has digits that occur exactly once each

There are no featured numbers above 9,999,999,999 (the next integer will have 11 digits and will have to repeat a digit)

---

**Explicit Requirements:**
- Return the next feature number greater than the integer argument
- Return an error message if there is no next featured number.

---


**Implicit Requirements:**
- Method should be named #featured
---

**Clarifying Questions:**

1. Can the input integer be 0 or negative?
  No, assume a positive integer input.

---

**Mental Model:**

#featured
Enter a loop.
  Increment the input integer by 1.
  Check if the result is an odd number, is a multiple of 7, and has digits that occur exactly once.
  If so break out of the loop and return the number.
  If the number is greater than 9_876_543_210, then break out of the loop and return an error message.
  
#uniq_digits
Convert the number into an array of digits. Check if the number of unique digits equals the number of digits in the array.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

featured(12) == 21
featured(20) == 21
featured(21) == 35
featured(997) == 1029
featured(1029) == 1043
featured(999_999) == 1_023_547
featured(999_999_987) == 1_023_456_987

featured(9_999_999_999) # -> There is no possible number that fulfills those requirements

---

Data Structure
--------------
Nothing complex. Use arrays to manage the digits.
---

Algorithm
---------

#featured
- Initialize variable 'possible_featured_number' to the input integer
- Enter a loop
  - Increment 'possible_featured_number' by 1.
  - Return 'possible_feature_number' if 'posible_feature_number' is:
    - odd AND
    - a multiple of 7 AND
    - has no repeating digits
  - Return an error message if 'possible_feature_number' is greater than 9_876_543_210

Code
----

=end

ERROR = 'There is no possible number that fulfills those requirements.'

def uniq_digit?(num)
  digits = num.to_s.split('')
  digits.size == digits.uniq.size
end

def featured_number?(num)
  (num.odd?) && (num % 7 == 0) && (uniq_digit?(num))
end

def featured(num)
  possible_featured_number = num
  loop do
    possible_featured_number += 1
    return possible_featured_number if featured_number?(possible_featured_number)
    return ERROR if possible_featured_number > 9_876_543_210
  end
end

p featured(12) == 21
p featured(20) == 21
p featured(21) == 35
p featured(997) == 1029
p featured(1029) == 1043
p featured(999_999) == 1_023_547
p featured(999_999_987) == 1_023_456_987

p featured(9_999_999_999) # -> There is no possible number that fulfills those requirements

File: JS210_small_problems/medium_2/next_featured_number_higher_than_a_given_value.js
"use strict";
/*

A featured number (something unique to this exercise) is an odd number that is a multiple of 7, with all of its digits occurring exactly once each. For example, 49 is a featured number, but 98 is not (it is not odd), 97 is not (it is not a multiple of 7), and 133 is not (the digit 3 appears twice).

Write a function that takes an integer as an argument and returns the next featured number greater than the integer. Issue an error message if there is no next featured number.

NOTE: The largest possible featured number is 9876543201.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: integer argument
-  Output: the next featured number higher than the integer argument

---

**Problem Domain:**

A featured number (something unique to this exercise) is an odd number that is a multiple of 7, with all of its digits occurring exactly once each.

The largest featured number is 9876543201.

---

**Explicit Requirements:**
- If there is a featured number greater than the input integer, return the next featured number.
- If the input is greater than or equal to 9876543201, issue an error message - "There is no possible number that fulfills those requirements."
- A featured number is an odd number that's a multiple of 7, will all of its digits occurring exactly once.
---


**Implicit Requirements:**
- We should return the next largest featured number (if it exists) if the input is also a featured number.
---

**Clarifying Questions:**

1. When issuing the error message, should it be returned or logged to the console?
  Logged to the console.
2. Will the input always be an integer of the Number type?
  Yes.
3. Will the input always be a positive integer?
  Yes.
4. Will there always be a single argument, no more and no fewer?
  Yes.
  
---

**Mental Model:**

- Check if the input is greater than or equal to the largest featured number 9876543201.
  - If so, log the error message.
- Find the next odd, multiple of seven that's greater than the input integer.
- Loop
  - If the current number is a featured number (at this point, just making sure all the digits occur once), return the current number.
  - Increment the current number by 14.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

featured(12);           // 21
featured(20);           // 21
featured(21);           // 35
featured(997);          // 1029
featured(1029);         // 1043
featured(999999);       // 1023547
featured(999999987);    // 1023456987
featured(9876543186);   // 9876543201
featured(9876543200);   // 9876543201
featured(9876543201);   // "There is no possible number that fulfills those requirements."

---

Data Structure
--------------

Use numbers. 

Use strings and arrays when determining if the digits in a number only occur once.

---

Algorithm
---------

#featured
- Check if the input is greater than or equal to the largest featured number 9876543201.
  - If so, log the error message.
- Find the next odd, multiple of seven that's greater than the input integer with #nextOddMultipleOf7
- Loop
  - If the current number has only unique digits (checked with #hasUniqueDigits), return the current number.
  - Increment the current number by 14.

#nextOddMultipleOf7(number)
- Loop
  - increment number by 1
  - if number is odd AND is a multiple of 7, return it

#hasUniqueDigits(number)
- convert the number into a string
- split the number into an array of digits
- sort the array of digits
- iterate over the array of digits
  - if the current element is the same as the next element in the array, return false
- return true if iteration completes without returning early

Code
----

*/

function nextOddMultipleOf7(number) {
  while(true) {
    number += 1;
    if (number % 2 === 1 && number % 7 === 0) return number;
  }
}

function hasUniqueDigits(number) {
  let digits = String(number).split('');

  let seenDigits = {};

  for (let index = 0; index < digits.length; index += 1) {
    if (seenDigits[digits[index]]) return false;

    seenDigits[digits[index]] = true;
  }

  return true;
}

function featured(number) {
  const MAX_FEATURED_NUM = 9876543201;

  if (number >= MAX_FEATURED_NUM) {
    console.log('There is no possible number that fulfills those requirements');
  } else {
    number = nextOddMultipleOf7(number);

    while (true) {
      if (hasUniqueDigits(number)) return number;
      number += 14;
    }
  }
}

console.log(featured(12));           // 21
console.log(featured(20));           // 21
console.log(featured(21));           // 35
console.log(featured(997));          // 1029
console.log(featured(1029));         // 1043
console.log(featured(999999));       // 1023547
console.log(featured(999999987));    // 1023456987
console.log(featured(9876543186));   // 9876543201
console.log(featured(9876543200));   // 9876543201
console.log(featured(9876543201));   // "There is no possible number that fulfills those requirements."

// LS Solution

// function featured(number) {
//   const MAX_FEATURED = 9876543201;
//   let featuredNum = toOddMultipleOf7(number);

//   do {
//     if (allUnique(featuredNum)) {
//       return featuredNum;
//     }

//     featuredNum += 14;
//   } while (featuredNum <= MAX_FEATURED);

//   return 'There is no possible number that fulfills those requirements.';
// }

// function toOddMultipleOf7(number) {
//   do {
//     number += 1;
//   } while (number % 2 === 0 || number % 7 !== 0);

//   return number;
// }

// function allUnique(number) {
//   let digits = String(number).split('');
//   let seen = {};

//   for (let idx = 0; idx < digits.length; idx += 1) {
//     if (seen[digits[idx]]) {
//       return false;
//     }

//     seen[digits[idx]] = true;
//   }

//   return true;
// }